File: mockspresso-dagger/src/main/java/com/episode6/hackit/mockspresso/dagger/DaggerLazyMaker.java
package com.episode6.hackit.mockspresso.dagger;

import com.episode6.hackit.mockspresso.api.DependencyProvider;
import com.episode6.hackit.mockspresso.api.SpecialObjectMaker;
import com.episode6.hackit.mockspresso.reflect.DependencyKey;
import com.episode6.hackit.mockspresso.reflect.TypeToken;
import dagger.Lazy;

import java.lang.reflect.ParameterizedType;
import java.lang.reflect.Type;

/**
 * An implementation of SpecialObjectMaker for {@link dagger.Lazy}.
 */
public class DaggerLazyMaker implements SpecialObjectMaker {

  @Override
  public boolean canMakeObject(DependencyKey<?> key) {
    return key.typeToken.getRawType() == Lazy.class &&
        key.typeToken.getType() instanceof ParameterizedType;
  }

  @SuppressWarnings("unchecked")
  @Override
  public <T> T makeObject(DependencyProvider dependencyProvider, DependencyKey<T> key) {
    if (!canMakeObject(key)) {
      return null;
    }

    Type paramType = ((ParameterizedType)key.typeToken.getType()).getActualTypeArguments()[0];
    DependencyKey<?> subKey = DependencyKey.of(TypeToken.of(paramType), key.identityAnnotation);
    return (T) new DependencyInstanceLazy(dependencyProvider, subKey);
  }

  private static class DependencyInstanceLazy<V> implements Lazy<V> {
    private final DependencyProvider mDependencyProvider;
    private final DependencyKey<V> mDependencyKey;

    private DependencyInstanceLazy(
        DependencyProvider dependencyProvider,
        DependencyKey<V> dependencyKey) {
      mDependencyProvider = dependencyProvider;
      mDependencyKey = dependencyKey;
    }

    @Override
    public V get() {
      return mDependencyProvider.get(mDependencyKey);
    }
  }
}

File: mockspresso-dagger/src/main/java/com/episode6/hackit/mockspresso/dagger/DaggerLazyMaker.kt
package com.episode6.hackit.mockspresso.dagger

import com.episode6.hackit.mockspresso.api.DependencyProvider
import com.episode6.hackit.mockspresso.api.SpecialObjectMaker
import com.episode6.hackit.mockspresso.reflect.DependencyKey
import com.episode6.hackit.mockspresso.reflect.genericParameterKey
import com.episode6.hackit.mockspresso.reflect.isParameterizedGeneric
import dagger.Lazy

/**
 * An implementation of SpecialObjectMaker for [dagger.Lazy].
 */
internal class DaggerLazyMaker : SpecialObjectMaker {
  override fun canMakeObject(key: DependencyKey<*>): Boolean = key.isParameterizedGeneric(Lazy::class)
  override fun makeObject(dependencyProvider: DependencyProvider, key: DependencyKey<*>): Any? {
    if (!canMakeObject(key)) return null
    return Lazy { dependencyProvider[key.genericParameterKey()!!] }
  }
}